ssz/src/main/java/tech/pegasys/teku/ssz/backing/cache/SoftRefIntCache.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.ssz.backing.cache;
+
+import java.lang.ref.SoftReference;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Supplier;
+
+/**
+ * Keeps the delegate cache in a {@link SoftReference} to allow the cache to be GC'ed if the
+ * application lacks of heap memory.
+ *
+ * <p>On {@link #copy()} or {@link #transfer()} also returns a {@link SoftRefIntCache} instance
+ */
+public class SoftRefIntCache<V> implements IntCache<V> {
+
+  private final Supplier<IntCache<V>> cacheCtor;
+  private volatile SoftReference<IntCache<V>> delegate;
+
+  private SoftRefIntCache(IntCache<V> initialDelegate, Supplier<IntCache<V>> cacheCtor) {
+    this.cacheCtor = cacheCtor;
+    delegate = new SoftReference<>(initialDelegate);
+  }
+
+  public SoftRefIntCache(Supplier<IntCache<V>> cacheCtor) {
+    this(cacheCtor.get(), cacheCtor);
+  }
+
+  public IntCache<V> getDelegate() {
+    IntCache<V> cache = delegate.get();
+    if (cache == null) {
+      cache = cacheCtor.get();
+      delegate = new SoftReference<>(cache);
+    }
+    return cache;
```

Comment:
I'm assuming we're ok with multiple threads winding up creating new caches at the same time here?  One of them will just wind up being GC'd again pretty much immediately which is the same behaviour we'd see if a single new `SoftReference` as created and the GC immediately cleared it out again.

Reply:
Yes, concurrent access case that should work fine here. It can lead to an additional cache miss in that case but that trade off seems OK to me. 
I decided to avoid `synchronized` here as concurrent access would not be the usual case IMHO